### PR TITLE
Allow balance payments for any attendee status

### PR DESCRIPTION
**Pull request by stefan-burke**
## What changed

Customers can now pay an outstanding balance online no matter which attendee status they are in. The old reservation-only gate is gone from the public `/pay` flow and from the admin ledger panel.

The refund and cleanup paths were tightened at the same time:

- Balance payments store provider references for later refunds.
- Admin refunds include both new processed payment references and legacy `payment_id` charges.
- The refresh-payment route loads every refund reference (legacy `payment_id` plus stored `processed_payments` rows), asks the provider about each one, and only records the ledger refund when every charge is confirmed refunded at the provider — so a balance charge already refunded directly in the provider dashboard is caught alongside the deposit rather than failing with `refund_not_recorded`.
- The refund ledger only records a full refund when the refunded provider references cover the ledger money being reversed.
- Manual payments, write-offs, and uncovered payment groups fail closed for manual adjustment.
- Merged attendees keep their processed payment references **and** their source's legacy `payment_id` (repointed onto the target as a `processed_payments` row), so pre-migration source charges stay refundable after the source attendee is deleted.
- Balance-paid attendees with stored provider references are not shown or deleted as incomplete payments. Refunded attendees whose stored reference gets pruned (because a `refund_cash` leg now exists) are also correctly excluded — the predicate reads the ledger-fed `refunded` flag rather than the now-pruned reference.
- Balance settlement only moves reservation statuses to the paid default; owner-set non-reservation statuses are preserved.

Follow-up cleanup:

- Removed the obsolete balance-payment refund plan document.
- Split the listing page management panel loaders out of the read-only listing page data loader.
- Split refund state, UI, and provider-log tests into separate files under the project size target.

## What this means

- Customers with real money owed can pay from their link regardless of status.
- Owners can refund balance-paid attendees from the app, including old deposit plus new balance-charge combinations, and refresh-payment reports an accurate status when balances have been refunded in the provider dashboard.
- Merged attendees keep refund access to both the target's and the source's charges, even when one of those was created before this PR added the new reference column.
- Admin cleanup no longer treats successfully balance-paid attendees as failed payments just because their legacy `payment_id` is empty, and keeps refunded attendees out of the failed-payments list even after their stored reference is pruned.
- Manual/operator money stays a manual adjustment decision instead of being silently reversed as a provider refund.
- The touched source and test files now stay below the preferred ~400-line limit.

## Checks

- `deno task lint:ci`
- `deno task typecheck`
- `deno task cpd`
- `deno task build:edge`
- `deno task test:coverage` (`15513 passed`; final coverage check reported 100% line and branch coverage)
- Mutation tests were not run, per request.

## Summary by CodeRabbit

* **New Features**
  * Admin refunds now use encrypted payment references and support both balance-payment and legacy charges.
  * Added a provider-refund refresh flow and improved bulk refund-all handling with clearer success/failure outcomes.
  * Attendee merges preserve the source’s payment reference needed for continued refund processing.
  * Public outstanding balances can now be paid for eligible non-reservation attendees.
* **Bug Fixes**
  * Refund eligibility and visibility now correctly reflect active booking lines, payment references, and already-refunded state.
  * Listing overview/incomplete-payment and attendee completion stats now account for recorded payment references.
  * Payment replay, refund status refresh, and refund ledger recording are more robust and consistent.

### CODE REVIEW

File: src/features/public/balance.ts [hunk 1]
```diff
@@ -49,10 +47,9 @@ const withOutstanding = async (
   const payload = await verifyBalanceToken(token);
   const state = payload ? await getAttendeeBalanceState(payload.a) : null;
   if (!payload || !state) return htmlResponse(balanceInvalidPage());
-  const status = state.statusId
-    ? await getAttendeeStatus(state.statusId)
-    : null;
-  if (!status?.is_reservation || state.remainingBalance <= 0) {
+  // Any attendee who still owes money can pay it online — the balance to
+  // collect is what's outstanding, whatever status the booking sits in.
+  if (state.remainingBalance <= 0) {
```
[the remark this answers is not in the document]
Comment (stefan-burke) [reply src/features/public/balance.ts:52]: Not actionable — no fee is under-collected. Provider-less bookings (`ticket-submit.ts` recasts with `feeSubtotal: 0` and posts **no `fee` leg** and no `payment` leg) and admin-set balances (`attendee-form-model.ts` `toLedgerOrder`) never book a booking fee, so none is part of the owed balance; the remaining balance is purely gross minus payments. A reservation pays its whole fee up front at deposit time (`checkout-pricing.ts` `priceCheckout`). So `feeSubtotal: 0` on the balance checkout collects exactly what is outstanding in every case — reservation and non-reservation alike.

---
_Generated by [Claude Code](https://claude.ai/code)_

[the remark this answers is not in the document]
Comment (stefan-burke) [reply src/features/public/balance.ts:52]: Real, but pre-existing and not introduced by this change. `getAttendeeOrderSummary` derives `depositPaid` from `pricePaidFromLedger`, which sums the gross `sale` leg (not cash received), and `fullPrice = depositPaid + remainingBalance` then double-counts the unpaid remainder. That computation is shared and unchanged — it already produced these same overstated totals for reservation balances, and it's documented as an accepted quirk in `test/lib/server-reservation/edge-cases.test.ts:63-71`. Removing the reservation restriction only widened *who reaches the page*, not the numbers. Tracked as a separate follow-up (see §8 of `docs/balance-payment-refund-rework.md`), since fixing it means changing the summary to report cash received rather than the gross sale leg.

---
_Generated by [Claude Code](https://claude.ai/code)_

[the remark this answers is not in the document]
Comment (stefan-burke) [reply src/features/public/balance.ts:52]: Confirmed — real, and newly reachable by this change. A non-reservation attendee with an empty `payment_id` (provider-less booking or admin-added) can now pay a balance online but can't be refunded in-app, because the refund routes reject an empty `payment_id`. Digging in, it's part of a broader gap: even for reservations, the balance charge's provider reference is never persisted (`processed_payments` stores only the session id), and `soleBookingOrder` punts every balance-settled account (two event groups) to manual adjustment — so balance charges aren't cleanly refundable in-app for anyone today. We're reworking this properly rather than papering over the guard. Design committed for review in `docs/balance-payment-refund-rework.md` (commit ec41202): persist every provider charge's reference, refund them all, and reverse the whole ledger account so the `refunded` flag sets. Implementation follows once the plan is reviewed.

---
_Generated by [Claude Code](https://claude.ai/code)_

[the remark this answers is not in the document]
Comment (stefan-burke) [reply src/features/public/balance.ts:52]: Good catch — real and newly reachable. One wrinkle worth flagging: that `UPDATE attendees SET status_id …` statement is also the settle's **verdict** — its `rowsAffected === 0` is how `settleAttendeeBalance` detects an amount mismatch (balance changed mid-checkout) and refunds. So simply skipping it for non-reservations would make every non-reservation settle report `amount_mismatch` and refund a valid payment. The fix has to split the verdict (always fires when `owed === expectedAmount`) from the paid-default move (gated on the current status being a reservation). Folded into the plan doc as decision §6.1 (commit 608cb0f), recommending exactly what you suggest — only auto-transition reservations, leave any owner-set status untouched — pending the maintainer's sign-off before I touch the settle path.

---
_Generated by [Claude Code](https://claude.ai/code)_

[the remark this answers is not in the document]
Comment (stefan-burke) [reply src/features/public/balance.ts:52]: Rechecked after the final implementation; no code change needed here. Provider-less and admin-added balances still do not book a fee leg, so there is no fee in the outstanding balance to collect later. The balance checkout using `feeSubtotal: 0` collects the money actually owed rather than waiving a recorded fee.

[the remark this answers is not in the document]
Comment (stefan-burke) [reply src/features/public/balance.ts:52]: Kept as a real but separate follow-up. The summary math is pre-existing and shared with reservation balances: it reports the gross sale leg as `depositPaid`, so fixing it means changing the order-summary model to show cash received. This PR leaves that behavior unchanged and focuses on opening the payment path plus making settlement/refunds safe.

[the remark this answers is not in the document]
Comment (stefan-burke) [reply src/features/public/balance.ts:52]: Addressed in the pushed branch (`d7c57d7`). Balance payments now persist provider references in `processed_payments`, and admin refunds load those references even when the legacy `attendee.payment_id` is empty. Regression coverage includes balance-payment refund routes plus `test/shared/db/payment-references.test.ts` and the refund-ledger tests.

[the remark this answers is not in the document]
Comment (stefan-burke) [reply src/features/public/balance.ts:52]: Addressed in the pushed branch (`d7c57d7`). Settlement now keeps the amount-match verdict separate from the status transition: the verdict still runs for every balance payment, but only reservation statuses move to the paid default. Owner-set non-reservation statuses are preserved after payment.

File: src/shared/db/payment-references.ts [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+    }
+  }
+  return new Map(
+    attendees.map((attendee) => {
+      const references = asRefundReferences(byAttendee.get(attendee.id)!);
+      return [
+        attendee.id,
+        references.length === 0 && attendee.payment_id !== ""
+          ? [legacyReference(attendee.payment_id)]
+          : references,
```
[the remark this answers is not in the document]
Comment (stefan-burke) [reply src/shared/db/payment-references.ts]: Addressed in the pushed branch (`d7c57d7`). `getRefundPaymentReferences` now appends the legacy `payment_id` whenever it is not already present in processed references, so a deposit plus later balance charge is refunded as one complete set. Locked by `test/shared/db/payment-references.test.ts`.

File: src/shared/refund-ledger.ts [hunk 3]
```diff
@@ -29,30 +27,20 @@ import type { Transfer, TransferInput } from "#shared/ledger/types.ts";
...
-  const groups = groupBy(legs, (leg) => leg.eventGroup);
-  if (groups.size !== 1) return null;
-  const order = [...groups.values()][0]!;
-  return order.some((leg) => recognisesRevenue(leg.kind)) ? order : null;
-};
+export const accountRefundGroups = (legs: Transfer[]): Transfer[][] => [
+  ...groupBy(
+    legs.filter((leg) => !isRefundLeg(leg.kind)),
+    (leg) => leg.eventGroup,
+  ).values(),
```
[the remark this answers is not in the document]
Comment (stefan-burke) [reply src/shared/refund-ledger.ts:107]: Addressed in the pushed branch (`d7c57d7`). The refund ledger now checks the refunded provider references against the ledger event groups before reversing the account, and it fails closed for manual attendee payments, write-off corrections, and uncovered provider payment groups. Regression coverage is in `test/shared/refund-ledger.test.ts` and `test/shared/refund-ledger-batch.test.ts`.

File: src/shared/refund-ledger.ts [hunk 4]
```diff
@@ -29,30 +27,20 @@ import type { Transfer, TransferInput } from "#shared/ledger/types.ts";
...
  */
-export const soleBookingOrder = (legs: Transfer[]): Transfer[] | null => {
-  const groups = groupBy(legs, (leg) => leg.eventGroup);
-  if (groups.size !== 1) return null;
-  const order = [...groups.values()][0]!;
-  return order.some((leg) => recognisesRevenue(leg.kind)) ? order : null;
-};
+export const accountRefundGroups = (legs: Transfer[]): Transfer[][] => [
+  ...groupBy(
+    legs.filter((leg) => !isRefundLeg(leg.kind)),
```
[the remark this answers is not in the document]
Comment (stefan-burke) [reply src/shared/refund-ledger.ts:105]: Addressed in the pushed branch (`d7c57d7`). Attendee merge now repoints source `processed_payments` rows onto the target attendee instead of deleting them, so moved ledger groups keep their refundable provider references. Covered by `test/shared/merge/attendee-merge/repoint.test.ts`.

File: src/features/public/balance.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (stefan-burke) [reply src/features/public/balance.ts:52]: Addressed in the pushed branch (`d7c57d7`). Incomplete-payment classification now looks at real payment references, not just the legacy `payment_id`: roster stats, overview SQL stats, exports, and the delete-incomplete route all treat processed provider references as valid payment evidence. Covered by the listing template tests, overview stats tests, and `test/lib/server-attendees.test.ts`.

File: src/shared/incomplete-payment.ts [hunk 5]
```diff
@@ -0,0 +1,11 @@
...
+
+export const isIncompletePayment = (
+  attendee: Pick<Attendee, "price_paid" | "remaining_balance">,
+  hasPaidListing: boolean,
+  hasPaymentReference: boolean,
+): boolean =>
+  hasPaidListing &&
+  !hasPaymentReference &&
+  Number.parseInt(attendee.price_paid, 10) > 0 &&
+  attendee.remaining_balance <= 0;
```
[the remark this answers is not in the document]
Comment (stefan-burke) [reply src/shared/incomplete-payment.ts]: Addressed in 9c16c55. The predicate now requires `!attendee.refunded`, and `refunded` is the ledger-fed projection (`refundedFromLedger` in `src/shared/db/attendees/queries.ts:44`) that returns `true` iff a `refund_cash` leg sourced from this attendee exists. Once `prunePayments` deletes a processed_payment row it is because that `refund_cash` leg exists (see the prune clause in `src/shared/db/prune.ts:87-92`), so the same fact the prune used to delete the reference also flips `refunded` and excludes the attendee from the incomplete-payment classification — no separate non-pruned signal is needed. Locked by `returns false for refunded paid attendee with no surviving payment reference` in `test/templates/admin/listings/income-and-helpers.test.ts:274-281`, which constructs the exact scenario described (paid attendee, empty `payment_id`, `price_paid: "1000"`, no surviving payment reference, `refunded: true`) and asserts `isIncompletePayment` returns `false`.

File: src/shared/merge/attendee-merge.ts [hunk 6]
```diff
@@ -797,10 +797,12 @@ export const applyAttendeeMerge = async (
...
     ...insertStatements,
-    // Clean up source attendee
+    // Move source-owned payment references before deleting the source attendee,
+    // so refunds on the merged person can return every charge whose ledger rows
+    // now live on the target account.
     {
-      args: [sourceId],
-      sql: "DELETE FROM processed_payments WHERE attendee_id = ?",
+      args: [targetId, sourceId],
+      sql: "UPDATE processed_payments SET attendee_id = ? WHERE attendee_id = ?",
```
[the remark this answers is not in the document]
Comment (stefan-burke) [reply src/shared/merge/attendee-merge.ts:813]: Addressed in 9c16c55. The merge already receives `source.payment_id` as `sourcePaymentId` (`attendees-merge.ts:308`); the gap was that no statement wrote it onto the target. Added `legacyMergePaymentReferenceStatement` in `src/shared/db/payment-references.ts:208` — it inserts the source's legacy `payment_id` as a `processed_payments` row on the target keyed by `legacy-merge:<sourceId>` (a sentinel session id that `realSessionIds` deliberately strips out, so it travels only as a refundable charge reference and never re-matches against a booking event group). It runs inside the merge's atomic batch (line 807) alongside the `UPDATE processed_payments SET attendee_id = ? WHERE attendee_id = ?` repoint that moves the source's stored processed rows onto the target, so a pre-migration paid source's only charge reference is repointed (rather than deleted with the source row) and is reachable from `getRefundPaymentReferences` on the target afterwards. Locked by `preserves the source's legacy payment ID on the target` in `test/shared/merge/attendee-merge/repoint.test.ts:86-110`.

File: src/features/admin/attendees-edit.ts [hunk 7]
```diff
@@ -90,7 +90,9 @@ export const handleRefreshPayment: TypedRouteHandler<
 
     const isRefunded = await provider.isPaymentRefunded(attendee.payment_id);
     if (isRefunded && !attendee.refunded) {
-      const { posted } = await recordAttendeeRefund(attendeeId);
+      const { posted } = await recordAttendeeRefund(attendeeId, [
+        { sessionIds: [] },
+      ]);
```
[the remark this answers is not in the document]
Comment (stefan-burke) [reply src/features/admin/attendees-edit.ts]: Addressed in 9c16c55. The handler now loads every refund reference via `getRefundPaymentReferences([attendee])` (legacy `payment_id` plus every `processed_payments` row), refreshes each with `provider.isPaymentRefunded`, marks the already-refunded ones through `markPaymentReferencesProviderRefunded`, and only when every reference is provider-refunded does it call `recordAttendeeRefund(attendeeId, refreshedReferences)` — passing all references, so `coveredRefundGroups` (`refund-ledger.ts:85`) sees the balance charge's session id covering its provider-cash ledger group alongside the deposit's, instead of failing closed because the balance charge went unnoticed. Locked by two new tests in `test/lib/server-attendee-refresh-payment.test.ts`: `records provider-refunded deposit and balance charges` (both charges are queried at the provider and the ledger refund posts), and `reuses already-refunded balance charges before recording the refund` (an already-`provider_refunded_at`-marked balance reference is skipped rather than re-queried, and only the deposit's status is checked before recording).
